### PR TITLE
Michael Roger - Takeaway Challenge

### PR DESCRIPTION
**Pull request by Benaud12**
Changes made following first review feedback

### CODE REVIEW

File: spec/texting_twilio_spec.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+require 'texting_twilio'
+
+describe TextingTwilio do
+
+  describe '#send_text' do
+    it 'calls the correct request/response from Twilio API' do
+      VCR.use_cassette('twilio_send_text') do
+        response = TextingTwilio.send_text
+        expect(response.body).to include('Thank you! Your order has been placed and delivery is scheduled for')
```
Comment (tansaku) [inline spec/texting_twilio_spec.rb]: I would introduce an intermediate variable to reduce the line length here, e.g. 

```ruby 
expected_text = 'Thank you! Your order has been placed and delivery is scheduled for'
expect(response.body).to include(expected_text)
```

File: spec/restaurant_spec.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      expected_output = "1: Dishy1 | Price: £1.11\n"
+      expect(restaurant.view_menu).to eq expected_output
+    end
+
+    it 'lists multiple dishes with unique order numbers' do
+      restaurant.make_dish_available(dish1)
+      restaurant.make_dish_available(dish2)
+      restaurant.make_dish_available(dish3)
+      expected_output = "1: Dishy1 | Price: £1.11\n" +
+                        "2: Dishy2 | Price: £2.22\n" +
```
Comment (tansaku) [inline spec/restaurant_spec.rb]: why not just follow the suggestion, i.e.

```ruby
      expected_output = "1: Dishy1 | Price: £1.11\n" \
                            "2: Dishy2 | Price: £2.22\n" \
```

File: lib/restaurant.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  private
+
+  def no_dishes_available?
+    menu.count == 0
+  end
+
+  def valid_order_number?(number)
+    !!menu[number]
```
Comment (tansaku) [inline lib/restaurant.rb]: read the link ^^^

File: lib/restaurant.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    menu.each_with_index do |dish, index|
+      menu_string << "#{index+1}: #{dish.name} | Price: £#{'%.2f' % dish.price}\n"
+    end
+    menu_string
+  end
+
+  def select(dish_number)
+    fail "No dishes available on this menu" if no_dishes_available?
+    index = dish_number - 1
+    fail "There are no dishes matching the number given" unless valid_order_number?(index)
```
Comment (tansaku) [inline lib/restaurant.rb]: yeah 90 is probably okay - could extract error message to constant though ...

**Comment by tansaku** [comment]
hey @Benaud12 had a quick review - definitely improved.  Would like to see you fixing the rubocop and reek issues though.  At least for each one specifically tell me why you can't/won't fix:

```
$ rubocop --format simple
== lib/order.rb ==
C: 20:  3: Metrics/MethodLength: Method has too many lines. [6/5] (https://github.com/bbatsov/ruby-style-guide#short-methods)
C: 24: 81: Metrics/LineLength: Line is too long. [97/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)
== lib/restaurant.rb ==
C: 13:  3: Metrics/MethodLength: Method has too many lines. [6/5] (https://github.com/bbatsov/ruby-style-guide#short-methods)
C: 17: 67: Style/FormatString: Favor format over String#%. (https://github.com/bbatsov/ruby-style-guide#sprintf)
C: 17: 81: Metrics/LineLength: Line is too long. [82/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)
C: 25: 81: Metrics/LineLength: Line is too long. [90/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)
C: 36:  5: Style/DoubleNegation: Avoid the use of double negation (!!). (https://github.com/bbatsov/ruby-style-guide#no-bang-bang)
== lib/texting_twilio.rb ==
C:  7:  3: Metrics/MethodLength: Method has too many lines. [12/5] (https://github.com/bbatsov/ruby-style-guide#short-methods)
C: 15: 81: Metrics/LineLength: Line is too long. [118/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)
== spec/order_spec.rb ==
C: 34: 47: Style/LineEndConcatenation: Use \ instead of + or << to concatenate those strings.
C: 35: 48: Style/LineEndConcatenation: Use \ instead of + or << to concatenate those strings.
C: 46: 48: Style/LineEndConcatenation: Use \ instead of + or << to concatenate those strings.
C: 74: 40: Style/DotPosition: Place the . on the next line, together with the method name. (https://github.com/bbatsov/ruby-style-guide#consistent-multi-line-chains)
== spec/restaurant_spec.rb ==
C: 29: 54: Style/LineEndConcatenation: Use \ instead of + or << to concatenate those strings.
C: 30: 54: Style/LineEndConcatenation: Use \ instead of + or << to concatenate those strings.
== spec/texting_twilio_spec.rb ==
C:  9: 81: Metrics/LineLength: Line is too long. [111/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)
```

```
$ reek -U
lib/order.rb -- 2 warnings:
  [14, 16]:Order#add_to_basket calls @basket[dish] 2 times (DuplicateMethodCall) [https://github.com/troessner/reek/blob/master/docs/Duplicate-Method-Call.md]
  [31]:Order#submit is controlled by argument payment (ControlParameter) [https://github.com/troessner/reek/blob/master/docs/Control-Parameter.md]
lib/texting_twilio.rb -- 1 warning:
  [7]:TextingTwilio#self.send_text has approx 8 statements (TooManyStatements) [https://github.com/troessner/reek/blob/master/docs/Too-Many-Statements.md]
```

Great to see you using the VCR gem, but note that all your sensitive data is still in your `spec/vcr/twilio_send_text.yml` file.  You need to extract that for security.  The takeaway challenge rubric has the full details.

What I'd recommend is that you try and give yourself a full code review using the takeaway rubric doc?  Or maybe get another senior and do a peer review of each others code?  Make sure you've addressed all the issues and I can certainly give you a sticker.  For any issue that you can't/won't fix I'd like to see an explanation of why in the README ... :-)

Great work here - remember you can get a sticker at any time, but only if you tell me you really deserve it :-)  Do you think you deserve it, or is there a little more you could do here?

**Comment by Benaud12** [comment]
Thanks @tansaku . I think I may have misunderstood Hound on a couple of issues, so sorry for wasting your time with them. Will update as you've suggested and come back chasing that all important sticker....not quite sticker worthy just yet

**Comment by tansaku** [comment]
not a time waster @Benaud12 - always good to check in :-) and some of the rubocop/reek issues can potentially be dismissed but I want to make sure you are doing so for the right reason :-)

**Comment by Benaud12** [comment]
@tansaku - Thanks again for all your feedback on this. I've addressed all Hound, Reek and Rubocop  issues now, with the exception of 3 line length violations, but they're all below 90 and I think that's reasonable. (Also, Rubocop and Hound seem to disagree on where the '.' should be placed for multi-line method chaining, so I went with Hound)
I've just gone over my project using the takeaway rubric and feel it is now sticker worthy but would appreciate any further input you might have.
Thanks

**Pull request closed without merging**